### ISSUE TITLE
Some newbie questions on how to do things with libvips

### ISSUE THREAD

**Issue by augustocdias** [libvips/libvips#1449]
Background: I have a Rust web server that do some image processing using OpenCV and I want to switch to libvips mostly because of performance.

Lately I've been trying to write bindings for the libvips C library for rust and I'd like to start with the things I have to support in my application, which are decoding images from any known format, resizing, rotating, watermarking and encoding to jpeg or webp. 

I've been reading the docs and I've got a good idea on how the library works, but still got a couple doubts. Please bear in mind that I'm not a very experienced C programmer. I have some experience with C++, but still it's been almost 10 years since last time I used it.

Here it goes: 
1. For the functions that take varargs options (the option named parameters), how do I specify them? Are them strings and should be specified like the options in the read from file options where you can just pass in the file name the options like this `file.jpeg[shrink=2]`? So should I specify a list with of `option=value` as strings (const *char)? Still about this options, how should I specify the ones that are enums?

2. What are the types for the PNG options in the save function? https://libvips.github.io/libvips/API/current/VipsForeignSave.html#vips-pngsave
Is there a place where I can find the defaults for the various options when encoding and decoding images?

3. One of the features I have to support envolves removing the alpha channel of images when encoding to JPEG. I know it works out of the box, but I want a different behavior. The default way, it transform the transparent pixels to black, which makes sense, since 0 = transparent = black. How can I invert the alpha channel in a image with libvips to get the opposite result (the fully transparent pixels getting white instead of black)?

4. About the error handling... How should I proceed? Should I have a background thread consuming and clearing the error buffer from time to time? If so how would it be a reasonable time to do it without degrading the performance of my application (because of the locks)...

5. Still about error handling, if I call `vips_error_freeze`, the buffer will never be filled, meaning I never have to care clearing it, right?

6. For watermarking I found out I should use the `vips_composite2` function with the `VIPS_BLEND_MODE_OVERLAY` mode, but how can I set an opacity for the overlay image?

Thank you in advance.

**Comment by jcupitt** [comment]
Hi @augustocdias,

libvips has two C APIs. There's the obvious top one with functions like `vips_embed()`:

https://libvips.github.io/libvips/API/current/libvips-conversion.html#vips-embed

This makes heavy use of varargs and is supposed to be convenient for human C programmers to use. You write code like:

```C
                VipsImage *image;
                VipsImage *x;

                if( !(image = vips_image_new_from_file( input_filename, 
                        "access", VIPS_ACCESS_SEQUENTIAL,
                        NULL )) )
                        vips_error_exit( NULL );
                if( vips_embed( image, &x, 100, 100, 3000, 3000,
                        "extend", VIPS_EXTEND_MIRROR,
                        NULL ) )
                        vips_error_exit( NULL );
                g_object_unref( image );
                image = x;

                if( vips_image_write_to_file( image, output_filename, 
                        "Q", 90,
                        NULL ) )
                        vips_error_exit( NULL );
                g_object_unref( image );
```

However this is just horrible for people trying to write bindings. varargs is a complete pain.

Fortunately, this API is a very thin skin over the GObject core. Language bindings are supposed to target this lower layer instead. 

This GObject layer has several very nice features:

1. No varargs!! Woo!
2. Full run-time introspection. Bindings can explore the features of the libvips they are linked to at run-time and expose those features to users.
3. Small size. The core of the libvips Python binding, for example, is only a few 100 lines of code. For that, you get a full binding to every feature of the entire library.
4. Low maintenance. Since almost all the binding is generated (either at run-time for languages like Python or Ruby, or at compile-time for languages like C# and C++), you don't need to update the binding to support new libvips versions.

I would look at something like the C++ binding. It has a core here:

https://github.com/libvips/libvips/blob/master/cplusplus/VImage.cpp

The first ~400 lines define a method called `VImage::call` which will call any libvips operation from C++. You can write C++ like this:

```C++
VImage VImage::embed( int x, int y, int width, int height, VOption *options ) const
{
    VImage out;

    call( "embed",
        (options ? options : VImage::option())->
            set( "in", *this )->
            set( "out", &out )->
            set( "x", x )->
            set( "y", y )->
            set( "width", width )->
            set( "height", height ) );

    return( out );
}
```

To call `vips_embed()`. Then there's a thing called `gen-operators.py` that generates all these little wrapper functions:

https://github.com/libvips/libvips/blob/master/cplusplus/gen-operators.py

And that's it, really. 400 lines of C++ and a little Python for the entire binding.

**Comment by jcupitt** [comment]
On your specific questions:

1. It's a null-terminated list of name-value pairs, you probably guessed from the example above. But I would absolutely NOT use this API, it's not designed for language bindings. You can introspect enums and allow strings if you like (the Python binding does this) so you can write eg. `a = b.embed(10, 10, 1000, 1000, extend="mirror")`.
2. I'll add the missing types, thanks for pointing this out. You can find the types from the CLI, try entering `vips pngsave` at the command-line.
3. Set the `background` property on `jpegsave`, eg. (in Python) `a.write_to_file("a.jpg", background=255)`.
4. The Python binding grabs and clears the error log every time it creates an exception. It's not great, but probably the best you can do. The alternative would be to simply have a "I done bad" error return from your binding and let your users present the error log in whatever way they like.
5. Yes, freeze/thaw block and unblock the error log.
6. You need to scale the alpha by some constant.

**Comment by jcupitt** [comment]
I found this old libvips introspection demo, it might help:

https://gist.github.com/jcupitt/9da6abe9764365086b9c08a362f22f07

If you compile it, you can run it like this:

```
$ ./a.out
john@yingna ~/try $ ./a.out
vips operations:
VipsSystem (system), run an external command
VipsAdd (add), add two images
VipsSubtract (subtract), subtract two images
VipsMultiply (multiply), multiply two images
VipsDivide (divide), divide two images
... etc.
```

Or like this:

```
$ $ ./a.out embed
embed:
required input arguments:
	in
		Input
		Input image
		VipsImage
	x
		x
		Left edge of input in output
		int - min = -1000000000, max = 1000000000, default = 0
	y
		y
		Top edge of input in output
		int - min = -1000000000, max = 1000000000, default = 0
	width
		Width
		Image width in pixels
		int - min = 1, max = 1000000000, default = 1
	height
		Height
		Image height in pixels
		int - min = 1, max = 1000000000, default = 1
optional input arguments:
	extend
		Extend
		How to generate the extra pixels
		enum - values:
			0 - black (VIPS_EXTEND_BLACK)
			1 - copy (VIPS_EXTEND_COPY)
			2 - repeat (VIPS_EXTEND_REPEAT)
			3 - mirror (VIPS_EXTEND_MIRROR)
			4 - white (VIPS_EXTEND_WHITE)
			5 - background (VIPS_EXTEND_BACKGROUND)
		default = 0
	background
		Background
		Color for background pixels
		array of double
```

**Comment by augustocdias** [comment]
As rust supports calling variatic functions from C, I'm more tempted to try the higher level API (at least for now). Where can I find the docs for the low level one by the way?

And which function should I use to scale the alpha?

**Comment by jcupitt** [comment]
Just in the standard API docs, eg.:

https://libvips.github.io/libvips/API/current/VipsObject.html#vips-argument-map

That varadic API is not really higher-level, just one that's designed for human C programmers.

It's just `linear`:

https://libvips.github.io/libvips/API/current/libvips-arithmetic.html#vips-linear

It's what Python uses for image * constant. You can write eg. `a * [1, 1, 1, 0.2]` to multiply alpha by 0.2. 

Here's the C++ overload:

https://github.com/libvips/libvips/blob/master/cplusplus/VImage.cpp#L882

**Comment by augustocdias** [comment]
I think I rather deal with varargs than with function pointers when it comes to ffi...
I've just tried passing `"background", 255` to the optional varargs of `vips_jpegsave_buffer` and got a segmentation fault... The other parameters specified in the docs worked fine when I tried.

**Comment by jcupitt** [comment]
Try `vips` -- it's tell you all the types:

```
$ vips jpegsave_buffer
save image to jpeg buffer
usage:
   jpegsave_buffer in [--option-name option-value ...]
where:
   in           - Image to save, input VipsImage
outputs:
   buffer       - Buffer to save to, output VipsBlob
optional arguments:
   page-height  - Set page height for multipage save, input gint
			default: 0
			min: 0, max: 10000000
   Q            - Q factor, input gint
			default: 75
			min: 1, max: 100
   profile      - ICC profile to embed, input gchararray
   optimize-coding - Compute optimal Huffman coding tables, input gboolean
			default: false
   interlace    - Generate an interlaced (progressive) jpeg, input gboolean
			default: false
   no-subsample - Disable chroma subsample, input gboolean
			default: false
   trellis-quant - Apply trellis quantisation to each 8x8 block, input gboolean
			default: false
   overshoot-deringing - Apply overshooting to samples with extreme values, input gboolean
			default: false
   optimize-scans - Split the spectrum of DCT coefficients into separate scans, input gboolean
			default: false
   quant-table  - Use predefined quantization table with given index, input gint
			default: 0
			min: 0, max: 8
   strip        - Strip all metadata from image, input gboolean
			default: false
   background   - Background value, input VipsArrayDouble
operation flags: sequential nocache 
```

If you use the GObject API you'll get type safety for free, but of course it's your choice.

**Comment by jcupitt** [comment]
In C it would be (untested):

```C
VipsArrayDouble *bg = vips_array_double_newv(1, 255.0);
if (vips_jpegsave(image, filename, "background", bg, NULL))
    handle_error();
vips_area_unref(VIPS_AREA(bg));
```

Of course, if you use the GObject API, you get all the `_newv()` / `_unref()` boilerplate for free ...

**Comment by augustocdias** [comment]
I'll give it try
If I were using the GObject API I would have to write it like this?


```
   op = vips_operation_new( "jpegsave" );
    g_value_init( &gvalue, VIPS_TYPE_IMAGE );
    g_value_set_object( &gvalue, in );
    g_object_set_property( G_OBJECT( op ), "in", &gvalue );
    g_object_set_property( G_OBJECT( op ), "background", 255.0 );
    g_value_unset( &gvalue );
    g_object_unref( in ); 
```

**Comment by jcupitt** [comment]
I would do it like the C++ binding.

1. Write a general-purpose thing that can call *any* libvips operation.
2. A scrap of Python or whatever that automatically generates a set of convenient, typesafe wrappers.
3. Layer a few convenience functions over the top: operator overloads, for example.

The C# binding follows this pattern:

https://github.com/kleisauke/net-vips

If that's any help. The C++ binding is probably the closest to Rust though.

I suppose it depends on your aim: 

* if you only want to call a few functions and move on, of course do it in the simplest way you can.
* If your aim is to make a complete binding that will be maintained and used for a while, I think the GObject API is the one to target.

Another advantage of the gobject API is that it will generate the docs for you as well. The pyvips docs are mostly automatically generated, eg.:

https://libvips.github.io/pyvips/vimage.html#pyvips.Image.embed

**Comment by jcupitt** [comment]
I'd forgotten, I wrote a chapter for the docs abut all these issues:

https://libvips.github.io/libvips/API/current/binding.md.html

**Comment by augustocdias** [comment]
Yes
I based my last snippet on it 😅
Anyway 
Thank you for all the help

**Comment by augustocdias** [comment]
One last question that I think should apply to most image libraries...
Is it possible to actually increase the quality of a JPEG file?
I've checked with ImageMagick's identify command that I have a 75 quality JPEG... If I set the quality for this file to libvips encoder to 100, the file size will increase substantially...
Is the library actually enhancing the image or just spending more storage to store the same information?
And is it possible to get the image quality using the API?

**Comment by jcupitt** [comment]
No, there's no point increasing the Q beyond what it was originally saved as, unless you are somehow adding more detail -- perhaps drawing a PNG overlay, for example.

jpeg files don't record the Q setting inside, since it's actually a lot more complex than a single number. ImageMagick has some heuristics that it uses to try to guess the original Q, but it's only a guess and it's easy to trick it.

**Comment by augustocdias** [comment]
I am checking your introspection demo and I have a question. How to get the return type of an Operation? I'm trying to use it as a base for a generator... I thought about the python script, but that would input not only a python dependency but also pyvips... With the C one I depend only on the compiler and libvips (which has to be installed already anyway)

**Comment by jcupitt** [comment]
Hello again, all libvips operations have multiple inputs (some optional) and multiple outputs (some optional).

The C++ binding (for example) returns the first required output arg as the principal result, and returns the others through reference params. The optional inputs are given as a set of name-value pairs in a final argument (in a very ugly way ... C++11 would make this saner).

pyvips maps vips operations to something like

```python
a, b, c = operation(d, e, f, option1=x, option2=y)
```

Optional outputs are put in a hash as:

```python
a, b, c, opts = operation(d, e, f, option1=x, option2=y, enable_optional_output1=True)
print(f"opt out1 = {opts['optional_output1']}")
```

**Comment by augustocdias** [comment]
I mean the operations that returns 1 or 0 for sucess, failure...
The ones that are in params I managed to get.

**Comment by jcupitt** [comment]
Ah OK. Success/fail is the return value from `vips_operation_build()`.

**Comment by augustocdias** [comment]
Having this error when I try to use it:

```
gcc -o introspect -g -Wall introspect.c `pkg-config vips --cflags --libs`
introspect.c:180:17: warning: implicit declaration of function 'vips_operation_build' is invalid in C99 [-Wimplicit-function-declaration]
                vips_operation_build();
                ^
1 warning generated.
Undefined symbols for architecture x86_64:
  "_vips_operation_build", referenced from:
      _show_class in introspect-39a519.o
ld: symbol(s) not found for architecture x86_64
```

**Comment by augustocdias** [comment]
What I mean is
the C API has specific returns for successes and failures. The C++ API seems to not have as the output arguments are mapped to the return types of the functions...
I'd like to have something similar to the C API to be able to check when my calls fails or not instead of checking the error buffer.

**Comment by jcupitt** [comment]
Re. compile warning: sorry, of course I meant `vips_cache_operation_build()`. The VipsObject docs have a discussion on how building works:

https://libvips.github.io/libvips/API/current/VipsObject.html

And the operation cache layers on top of that:

https://libvips.github.io/libvips/API/current/VipsOperation.html#vips-cache-operation-build

Have a look at the C++ source to see exactly how to use it:

https://github.com/libvips/libvips/blob/master/cplusplus/VImage.cpp#L513

**Comment by jcupitt** [comment]
So the C++ API throws an exception if `vips_cache_operation_build()` fails. You'll want to map the libvips error return to something more rust-y, of course.

**Comment by augustocdias** [comment]
What is the correct way to re encode an image to a different format?
I'm trying to read it using the `vips_image_new_from_buffer` and then I call the foreign functions to save to the format I want (`webpsave_buffer_with_opts`, `pngsave_buffer_with_opts`, `jpegsave_buffer_with_opts` for example). For some tests I'm getting intermittent errors, meaning that if I keep retrying the same operation it eventually works.
I have a webserver that reads an image from one HTTP URL, get the byte array, initialise VipsImage with this array and re-encode them to the format I specify as parameter.
Another operation I make is composing images... I could make it run in a standalone application, but in my web server it doesn't work. I initialise VipsImage from buffer, resize the image, load another one also from buffer, compose both and in the end I call one of the foreign save to buffer to the respective format I want.

```
VIPS-WARNING **: 15:24:52.773: Not a PNG file
error buffer: vips_realpath: unable to form filename
unix error: No such file or directory
VipsJpeg: Not a JPEG file: starts with 0x10 0xc0
VipsJpeg: Not a JPEG file: starts with 0xed 0x1a
vips2png: unable to write to buffer
```

**Comment by jcupitt** [comment]
It should be very simple. If you are seeing intermittent problems they are probably caused by freeing something to early, or unreffing something too early.

Sample C code:

```C
#include <stdio.h>
#include <stdlib.h>

#include <vips/vips.h>

int
main( int argc, char **argv )
{
	char *buffer_in;
	size_t length_in;
	GError *error = NULL;
	VipsImage *image;
	char *buffer_out;
	size_t length_out;

	if( VIPS_INIT( argv[0] ) )
		vips_error_exit( NULL );

	/* Load source image into memory. This image will be in some format,
	 * like JPEG.
	 */
	if( !g_file_get_contents( argv[1], &buffer_in, &length_in, &error ) ) {
		fprintf( stderr, "unable to read file %s\n%s\n", 
			argv[1], error->message );
		g_error_free( error );
		exit( 1 );
	}

	/* Make a vips image from the memory buffer. This image will
	 * decompress from buffer_in as required.
	 */
	if( !(image = vips_image_new_from_buffer( buffer_in, length_in, NULL, 
		"access", VIPS_ACCESS_SEQUENTIAL,
		NULL )) ) 
		vips_error_exit( NULL );

	/* Create an output memory buffer. Again, this buffer will contain an
	 * image packed into some format, it will not be an array of pixel
	 * values.
	 */
	if( vips_image_write_to_buffer( image, 
		".jpg", (void **) &buffer_out, &length_out, NULL ) ) 
		vips_error_exit( NULL );

	if( !g_file_set_contents( argv[2], buffer_out, length_out, &error ) ) {
		fprintf( stderr, "unable to write file %s\n%s\n", 
			argv[1], error->message );
		g_error_free( error );
		exit( 1 );
	}

	g_object_unref( image ); 
	g_free( buffer_in );
	g_free( buffer_out );

	return( 0 );
}
```

Compile with:

```shell
gcc -g -Wall buffer.c `pkg-config vips --cflags --libs`
```

Run with:

```shell
./a.out k2.png x.jpg
```

**Comment by augustocdias** [comment]
One question: once I have a new instance of the VipsImage from an operation, like resize or composite2 for example, can I free the old one that was used as input?

**Comment by jcupitt** [comment]
Yes, operations always ref (or copy) all their inputs, and each output object refs the operation that made it. You can write code like this:

```C
VipsImage *im = vips_image_new ...;
VipsImage *x;

if( vips_invert( im, &x, NULL ) ) {
   g_object_unref( im );
   return( error code );
}
g_object_unref( im );
im = x;

if( vips_invert( im, &x, NULL ) ) {
   g_object_unref( im );
   return( error code );
}
g_object_unref( im );
im = x;

if( vips_invert( im, &x, NULL ) ) {
   g_object_unref( im );
   return( error code );
}
g_object_unref( im );
im = x;
```

So `im` always points to the head of the pipeline that's being built, and indirectly holds refs to everything earlier. You can free the whole thing by unreffing `im` once.

**Comment by augustocdias** [comment]
I think the problem was between moving the pointer between threads in my web app... I changed to do all the vips operations inside one single thread and it started working fine... I have to investigate why this is happening though...

Anyway... I've got a different problem when I'm building my app to run inside that [alpine image I've built](https://github.com/libvips/libvips/pull/1452). 

When I run the vips command, it works fine... But when I start my application it is failing on the `VIPS_INIT` call...
```
**
GLib-GObject:ERROR:../gobject/gtype.c:2743:g_type_register_static: assertion failed: (static_quark_type_flags)
```

Any idea why this could be happening? Could be that I'm linking my app incorrectly?
`ldd` output seems fine:

```
/lib/ld-musl-x86_64.so.1 (0x7f8ce00d8000)
	libvips.so.42 => /usr/local/lib/libvips.so.42 (0x7f8cdfbfa000)
	libgobject-2.0.so.0 => /usr/lib/libgobject-2.0.so.0 (0x7f8cdfbb0000)
	libssl.so.1.1 => /lib/libssl.so.1.1 (0x7f8cdfb31000)
	libcrypto.so.1.1 => /lib/libcrypto.so.1.1 (0x7f8cdf8b8000)
	libheif.so.1 => /usr/lib/libheif.so.1 (0x7f8cdf85e000)
	libpng16.so.16 => /usr/lib/libpng16.so.16 (0x7f8cdf82e000)
	libz.so.1 => /lib/libz.so.1 (0x7f8cdf814000)
	libimagequant.so.0 => /usr/lib/libimagequant.so.0 (0x7f8cdf802000)
	libtiff.so.5 => /usr/lib/libtiff.so.5 (0x7f8cdf798000)
	libjpeg.so.8 => /usr/lib/libjpeg.so.8 (0x7f8cdf702000)
	libgmodule-2.0.so.0 => /usr/lib/libgmodule-2.0.so.0 (0x7f8cdf6fd000)
	libexpat.so.1 => /usr/lib/libexpat.so.1 (0x7f8cdf6db000)
	libpangoft2-1.0.so.0 => /usr/lib/libpangoft2-1.0.so.0 (0x7f8cdf6c4000)
	libpango-1.0.so.0 => /usr/lib/libpango-1.0.so.0 (0x7f8cdf686000)
	libfontconfig.so.1 => /usr/lib/libfontconfig.so.1 (0x7f8cdf64a000)
	libgsf-1.so.114 => /usr/lib/libgsf-1.so.114 (0x7f8cdf608000)
	liborc-0.4.so.0 => /usr/lib/liborc-0.4.so.0 (0x7f8cdf59a000)
	liblcms2.so.2 => /usr/lib/liblcms2.so.2 (0x7f8cdf54a000)
	libgif.so.7 => /usr/lib/libgif.so.7 (0x7f8cdf53f000)
	librsvg-2.so.2 => /usr/lib/librsvg-2.so.2 (0x7f8cdf50d000)
	libglib-2.0.so.0 => /usr/lib/libglib-2.0.so.0 (0x7f8cdf413000)
	libintl.so.8 => /usr/lib/libintl.so.8 (0x7f8cdf403000)
	libcairo.so.2 => /usr/lib/libcairo.so.2 (0x7f8cdf316000)
	libwebpmux.so.3 => /usr/lib/libwebpmux.so.3 (0x7f8cdf30b000)
	libwebpdemux.so.2 => /usr/lib/libwebpdemux.so.2 (0x7f8cdf305000)
	libwebp.so.7 => /usr/lib/libwebp.so.7 (0x7f8cdf2b0000)
	libexif.so.12 => /usr/lib/libexif.so.12 (0x7f8cdf26f000)
	libc.musl-x86_64.so.1 => /lib/ld-musl-x86_64.so.1 (0x7f8ce00d8000)
	libffi.so.6 => /usr/lib/libffi.so.6 (0x7f8cdf266000)
	libde265.so.0 => /usr/lib/libde265.so.0 (0x7f8cdf1a3000)
	libx265.so.169 => /usr/lib/libx265.so.169 (0x7f8cdec12000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x7f8cdeabd000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x7f8cdeaa9000)
	libharfbuzz.so.0 => /usr/lib/libharfbuzz.so.0 (0x7f8cdea14000)
	libfreetype.so.6 => /usr/lib/libfreetype.so.6 (0x7f8cde962000)
	libfribidi.so.0 => /usr/lib/libfribidi.so.0 (0x7f8cde945000)
	libuuid.so.1 => /lib/libuuid.so.1 (0x7f8cde93c000)
	libgio-2.0.so.0 => /usr/lib/libgio-2.0.so.0 (0x7f8cde7b3000)
	libxml2.so.2 => /usr/lib/libxml2.so.2 (0x7f8cde68b000)
	libbz2.so.1 => /usr/lib/libbz2.so.1 (0x7f8cde67c000)
	libgdk_pixbuf-2.0.so.0 => /usr/lib/libgdk_pixbuf-2.0.so.0 (0x7f8cde65b000)
	libpangocairo-1.0.so.0 => /usr/lib/libpangocairo-1.0.so.0 (0x7f8cde64d000)
	libcroco-0.6.so.3 => /usr/lib/libcroco-0.6.so.3 (0x7f8cde617000)
	libpcre.so.1 => /usr/lib/libpcre.so.1 (0x7f8cde5ba000)
	libpixman-1.so.0 => /usr/lib/libpixman-1.so.0 (0x7f8cde529000)
	libxcb-shm.so.0 => /usr/lib/libxcb-shm.so.0 (0x7f8cde524000)
	libxcb.so.1 => /usr/lib/libxcb.so.1 (0x7f8cde4fd000)
	libxcb-render.so.0 => /usr/lib/libxcb-render.so.0 (0x7f8cde4ee000)
	libXrender.so.1 => /usr/lib/libXrender.so.1 (0x7f8cde4e2000)
	libX11.so.6 => /usr/lib/libX11.so.6 (0x7f8cde3c0000)
	libXext.so.6 => /usr/lib/libXext.so.6 (0x7f8cde3ad000)
	libgraphite2.so.3 => /usr/lib/libgraphite2.so.3 (0x7f8cde386000)
	libmount.so.1 => /lib/libmount.so.1 (0x7f8cde335000)
	libXau.so.6 => /usr/lib/libXau.so.6 (0x7f8cde330000)
	libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x7f8cde328000)
	libblkid.so.1 => /lib/libblkid.so.1 (0x7f8cde2de000)
	libbsd.so.0 => /usr/lib/libbsd.so.0 (0x7f8cde2c8000)
```

**Comment by augustocdias** [comment]
I believe I'll have to find a glibc instead of musl base image to work with...

**Comment by jcupitt** [comment]
Ah Alpine, yes, it has a tiny stack and it causes all kinds of random crashes. You have to be extremely cautious -- it's like living in a house with 1m50 ceilings.

**Comment by augustocdias** [comment]
In case you're interested
I've just published the bindings for libvips 
It doesn't cover everything, but most of it :)
https://crates.io/crates/libvips
Also an app that I'm using it for 
https://github.com/augustocdias/rustbier

Thank you for all the help.

**Comment by jcupitt** [comment]
Huh that looks very nice!

I guess you've seen the true streaming libvips branch? That would be interesting to test with your S3 setup.

https://github.com/libvips/libvips/pull/1443

**Comment by augustocdias** [comment]
Woah. That looks awesome. I’ll give it a try later

**Comment by augustocdias** [comment]
And by the way
could you give a look on this function to see if I'm doing it the most efficient way?
https://github.com/augustocdias/rustbier/blob/master/src/image_processor/mod.rs#L6

**Comment by jcupitt** [comment]
Hi, I had a quick look:

1. Use `thumbnail_buffer`

You are doing `new_from_buffer` then `resize`. You'll find the combined `thumbnail_buffer` operation is much faster (perhaps 4x faster) since it can exploit loader features like shrink-on-load. It's also a sequential operation, so it can overlap decompress and resize.

It also handles premultiplication for you, so you should see a good improvement in quality for transparent images.

2. Should `watermaked` be `rotated`?

Just a typo I guess.

```rust
    debug!("Rotating image to {:?}", rotation);
    let mut watermarked = if let Some(rotation) = rotation {
```

3. Render to memory before rotate

`thumbnail` is a sequential operation, so if you want to rotate, you'll have to render to memory beforehand. Don't do this is there's no rotation.

4. Adjust alpha before upsize

I think the watermark can be upsized before having its alpha adjusted, is that right? I would adjust alpha first to reduce computation (you'll be adjusting the alpha on fewer pixels).

5. Colour management and strip

I would consider adding colour management too. You can make images a LOT smaller (in many cases) by converting to sRGB and stripping metadata.

6. Support `reduction-effort`

The most useful control for webpsave is actually `reduction-effort` -- I'd consider supporting that.

**Comment by augustocdias** [comment]
1. woah. I'll try that one
2. I should rename this variable... It's misleading
3. I have to rotate the main image only... So I should call `thumbnail_buffer` and then `rotate` right?
4. Should I use `thumbnail_buffer` to load the watermarks already with the right size as well or open then, apply the alpha and then resize?
5. This is just passing the correct ICC profile right? I set it to default to `sRGB` always
6. Thanks. I'll try that :)

**Comment by jcupitt** [comment]
`thumbnail` works really well for large jpg images:

```
$ time vips thumbnail wtc.jpg x.jpg 128
real	0m0.248s
user	0m0.225s
sys	0m0.024s
$ time vips resize wtc.jpg x.jpg $(bc <<< "scale=10; 128 / $(vipsheader -f width wtc.jpg)")
real	0m1.186s
user	0m1.130s
sys	0m0.045s
```

So 5x (!!) faster there.

3.) Yes, `thumbnail_buffer` then `rotate`, but you will need to copy the output of `thumbnail_buffer` to memory first or you'll break streaming. Python has `.copy_memory()` for this, or in C it's

```C
VipsImage *vips_image_copy_memory( VipsImage *image );
```

It looks like a copy operation, but behind the screens will render a delayed computation to an area of memory.

On 4), no idea! It probably doesn't make much difference, but you could experiment.

5.) libvips has a built-in high-quality srgb profile, so I would just pass the string `"srgb"` to `thumbnail_buffer` for `export_profile`. All savers have a `strip` option to save without metadata.

**Comment by augustocdias** [comment]
About number 3: the `thumbnail_buffer` gives me already a `VipsImage` object as output... Can't I just use it as input for the `rotate`? Why do I need to copy?

**Comment by jcupitt** [comment]
`thumbnail` is a streaming operation. It doesn't compute anything immediately, instead it makes a pipeline where, if you pull on the end, it'll pull pixels from the source image in the same order.

If you add a 90 degree rotate to the pipe, ordering will break when you send the result to the output file. The writer will pull pixels left-to-right and top-to-bottom, but the 90 degree rotate will turn that into top-to-bottom then left-to-right on the input image. Inputs like JPG only support top down reading, so you'll get an error.

If you add the `copy_memory`, it'll run the `thumbnail` step into the memory area, then rotate from that to the output.

**Comment by augustocdias** [comment]
I was trying the thumbnail function... I think I won't be able to use due to business logic rules... It's not always that I'll get a target width or height to resize, so it might happens that the user request just a target height for example, which then I'd have to calculate what would be the width to keep the same proportion (that is also a requirement, images cannot lose their proportions)

Anyway... I was checking what would be the bottleneck of in my code and looks like it is reencoding the images:

```READ FILES: 99
READING FROM BUFFER: 0
RESIZING IMAGE: 2
ROTATING IMAGE: 2
WATERMARKING IMAGE: 5
REENCODING IMAGE: 246```

Times in milis...

Is there a way to optimise it inside libvips?

**Comment by jcupitt** [comment]
You can do constrained-width-but-unconstrained-height like this:

```ruby
a = Vips::thumbnail_buffer my_buffer, Vips::MAX_COORD, height: 128
```

ie. resize to fit within a very wide (10000000 pixels) but shallow rect.

Re. your benchmarks: yes, that's expected, libvips is demand-driven, so all computation is delayed until you do the final write to file. That 246ms is every stage executing.

If you want to time just reencode, swap the final `write_to_file` for `write_to_memory`, time that and subtract. You'll find jpg write is pretty quick, tif write is fairly quick, and png write takes 100s of times longer and no one knows why.

**Comment by augustocdias** [comment]
I am actually calling the foreign `save_buffer` variants and in indeed the Jpeg is way faster than the others... But it is still clearly the one operation that takes more time (understandable). 
I was just wondering if there was any config or parameter to make it faster :) 

About the thumbnail_buffer, the width is a required parameter... What happens if I send 0 or a negative number? Does it ignores it?

Also about the thumbnail_buffer with the image_copy_memory... should I unref the input parameter from the copy_memory operation? I guess it would return the same reference right (meaning I should call unref just once after I finish working with the result of this call)?

**Comment by jcupitt** [comment]
`save_buffer` will write a formatted image and will use the format library. If you use `write_to_memory` it'll simply write a large RGBRGB etc. array, which will take (effectively) zero time.

Try swapping `save_buffer` for `write_to_memory` and see how the timing changes.

I guess you are using libjpeg-turbo? Otherwise, that's probably it, sorry :( 

Re. thumbnail: just set the width to a very large number.

`copy_memory` is exactly like `copy`, so it returns an image with a ref that you own, and you can unref the input afterwards.

**Comment by jcupitt** [comment]
Here's another way to benchmark. This uses `avg` (average pixel value) -- it's fast to calculate, but will force the computation of every earlier stage in the pipeline.

```python
#!/usr/bin/env python3

import sys
import time
import pyvips

def took(msg, start):
    print(f"{msg} - {(time.time() - start) * 1000:.2f}ms")

def build_pipeline(filename):
    image = pyvips.Image.thumbnail(filename, 1024)
    watermark = pyvips.Image.new_from_file("watermark.png", access="sequential")
    watermark *= [1, 1, 1, 0.6]
    return image.composite(watermark, "over", x=100, y=100)

start = time.time()
image = build_pipeline(sys.argv[1])
result = image.avg()
took("avg", start)

start = time.time()
image = build_pipeline(sys.argv[1])
image.write_to_file("x.jpg")
took("jpg save", start)

start = time.time()
image = build_pipeline(sys.argv[1]) 
image.write_to_file("x.png")
took("png save", start)
```

I see:

```
$ ./bench.py ~/pics/k2.jpg
avg - 188.70ms
jpg save - 204.19ms
png save - 434.61ms
```

So load/resize/watermark takes 188ms, adding jpg save adds just 16ms (libvips will overlap save with load, so you get it for free if it's fast enough), but awful png save adds 246ms.

**Comment by augustocdias** [comment]
That's interesting... And a bit weird... why it is taking so long in my machine...
Might be the libjpeg as well... I've installed in my machine through `brew`, so I don't know which dependencies it used alongside. 
Only in my docker image I did the installation manually and with the specific dependencies I wanted.

Still, I believe the timing for resizing and watermarking even without using thumbnail was pretty impressive... And I guess is understandable that it takes long to write the buffer to a specific format as it has to go through all the image data to do so.

**Comment by jcupitt** [comment]
Ah, `brew` uses plain libjpeg, it's very hard to use libjpeg-turbo on MacOS unless you build the entire stack yourself. I've been trying to push them to move to libjpeg-turbo (it's ABI compatible) but no luck so far.

Writing to the buffer is quick -- it's encoding the image with the format library that's the slow part. The format libs are mostly single-threaded, and (especially libpng) do a lot of processing.

**Comment by augustocdias** [comment]
I have been checking the `imgproxy` application and we're both using the same functions basically. I've been running both under the same environment (alpine docker image with vips and jpeg turbo) and I'm seeing a difference of around 100ms on the same ops. When in my application only the encoding to jpeg is taking this this 100ms :(

**Comment by jcupitt** [comment]
Could that just be libjpeg-turbo?

**Comment by augustocdias** [comment]
Not sure... but it’s quite weird so different numbers on almost same operations

**Comment by jcupitt** [comment]
I think I'd make a tiny benchmark in pyvips and time that. It should be easy to work out what operations are being performed in what order.

**Comment by augustocdias** [comment]
I was playing with the `access` parameter on the read from buffer operation and sometimes I'm getting  this error when I use `SEQUENTIAL` access `VipsJpeg: out of order read at line 2648`
is there restriction on how I should use this?

**Comment by jcupitt** [comment]
Yes, that's the streaming stuff. You can only use a subset of libvips operations -- ones which will not do large-scale coordinate changes.

There's a chapter in the docs on this:

https://libvips.github.io/libvips/API/current/How-it-opens-files.md.html

**Comment by augustocdias** [comment]
I understand. 
And is there any performance gain in using the foreign loaders for each format instead of letting libvips find out which one to use? 
Maybe guessing it by the magic bytes

**Comment by jcupitt** [comment]
Things like rotate and vertical flip won't work. Most other things will.

You can check the operation flags, eg.:

```
$ vips invert
invert an image
usage:
   invert in out [--option-name option-value ...]
where:
   in           - Input image, input VipsImage
   out          - Output image, output VipsImage
operation flags: sequential 
```

libvips has a fast built in format finder based on things like magic bytes, so I'd just use that.

**Comment by augustocdias** [comment]
Nice... I've added a condition to use sequential when I don't have to rotate :)
I believe this will improve a bit. 
I believe part of my bottleneck is on IO, so I'll check if there's something I can do to fetch the images faster. 
Also, the number of threads. Vips starts its own thread pool and my app server also its own (and both taking the max available on the CPU as default)... I have to figure out who needs more threads... Do you have any advice regarding that? Giving more to vips will makes things better?

**Comment by jcupitt** [comment]
Hi again,

Yes, seq means that libvips can overlap image load, decode and processing (write is always overlapped). If load is slow (or if decode is slow) and processing is non-trivial, it should help.

By default, libvips makes as many threads as you have hardware threads. I'd leave it alone I think.

IMO the first step in benchmarking is to make a small test program that represents your work, and then try tuning the environment.

**Comment by augustocdias** [comment]
Why the addalpha is not in the vips operations list?
What is the best way to add alpha when an image doesn't have it?

I'm getting this error when using linear

```Vips error buffer: linear: vector must have 1 or 3 elements```

And I believe it is because the image doesn't have the alpha channel.

**Comment by jcupitt** [comment]
`addalpha` is just a convenience function for `bandjoin_const(255)`, ie. add a constant 255 band. Perhaps it should be a separate operation. `hasalpha?` is more complicated and has a C function.

**Comment by augustocdias** [comment]
cool. I'll use `bandjoin` then. And yes, I'm checking if the image has alpha :) 
Thanks for the all the help :) 
I haven't ever seen anyone so helpful like you in a OS project before.

**Comment by jcupitt** [comment]
Don't be too impressed -- I'm working on this stuff at the moment, so it's at the front of my mind. I'm not normally so responsive!

**Comment by cr-solutions** [comment]
Hi guys,

I like to port the libvips as native to PHP version 7.4 with FFI support.
Open an image as VipsImage and resizing looks like working, but if I like to save the image as file I get the following error:

"jpegsave: no property named `'

from function

"vips_image_write_to_file"

my FFI::cdef for this function looks like:

`int vips_image_write_to_file(VipsImage *image, const char *name);`

whats wrong here? What does this error mean?

**Comment by jcupitt** [comment]
Hi again @cr-solutions, please open a new issue rather than adding to old issues. It makes it much easier for people to search for answers if each issue is just about one question.
